README
/fire/project/python_project/jsend/README.md

changelog.md
version 0.7.1 released at 2014-07-25

====================================
changelog:
    fix a minor pychecker warning(Local variable (_j) not used)




version 0.7.0 released at 2014-07-13
====================================
changelog:
    fixed critical bug: new dict instance not dict reference.
    eg:
        # return a new dict instance(not reference) of dict(data)
        self['data'] = dict(data) if isinstance(data, dict) else data




version 0.6.0 released at 2014-07-12
====================================
changelog:
    make RSuccess, RFail, RError take default parameters in __init__ method.


version 0.5.1 released at 2012-08-09.
====================================
changelog:
    add a function parse_jsend(jsend_str)
        return a jsend object if @jsend_str is a valid jsend-json str else return original @jsend_str


version 0.5.0 2012-07-26 released
====================================
changelog: (2012-07-26)
In pracmatical usage, a key "message" for all responese is very convenient for development.
So add 'message' key to "Optional Keys" for type(success, fail)


    | ''Type'' | ''Description''                                        | ''Required Keys''  | ''Optional Keys'' |
    |----------+--------------------------------------------------------+--------------------+-------------------|
    | success  | All went well, and (usually) some data was returned.   | status, code, data | message           |
    |          |                                                        |                    |                   |
    | fail     | There was a problem with the data submitted,           | status, code, data | message           |
    |          | or some pre-condition of the API call wasn't satisfied |                    |                   |
    |          |                                                        |                    |                   |
    | error    | An error occurred in processing the request,           | status, message    | code, data        |
    |          | i.e. an exception was thrown                           |                    |                   |







version 0.4.0 2012-07-16 released
====================================
    fix: in pracmatical usage, code representing status is very
    import, so add 'code' key to "Required Keys" for "type" of
    "success", "fail". See the table below


    | ''Type'' | ''Description''                                        | ''Required Keys''  | ''Optional Keys'' |
    |----------+--------------------------------------------------------+--------------------+-------------------|
    | success  | All went well, and (usually) some data was returned.   | status, code, data |                   |
    |          |                                                        |                    |                   |
    | fail     | There was a problem with the data submitted,           | status, code, data |                   |
    |          | or some pre-condition of the API call wasn't satisfied |                    |                   |
    |          |                                                        |                    |                   |
    | error    | An error occurred in processing the request,           | status, message    | code, data        |
    |          | i.e. an exception was thrown                           |                    |                   |



version 0.3.0 2012-06-13 released
====================================
    fix: add setter for data(RSuccess, RFail, RError),
         add setter for code(RError)

version 0.2.4 2011-11-19 released
====================================

version 0.2.2 2011-11-18 released
====================================


jsend/test_jsend.py
# ** -- coding: utf-8 -- **
# !/usr/bin/env python
#
# Copyright (c) 2011 darkdarkfruit <darkdarkfruit@gmail.com>
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#

'''
jsend test
==========
You should have py.test installed first.
pip install -U pytest

Then, run
>> py.test  test_jsend.py



compatable with jsend:
http://labs.omniti.com/labs/jsend

shortely:
=================================================================
A basic JSend-compliant response is as simple as this:
{{{
{
    status : "success",
    data : {
        "post" : { "id" : 1, "title" : "A blog post", "body" : "Some useful content" }
     }
}
}}}

When setting up a JSON API, you'll have all kinds of different types of calls and responses.  JSend separates responses into some basic types, and defines required and optional keys for each type:

||''Type''||''Description''||''Required Keys''||''Optional Keys''||
||success ||All went well, and (usually) some data was returned.||status, data||||
||fail    ||There was a problem with the data submitted, or some pre-condition of the API call wasn't satisfied||status, data||||
||error   ||An error occurred in processing the request, i.e. an exception was thrown||status, message||code, data||
=================================================================

this module include 3 classes:
==============================
RSuccess --> Result Success
RFail    --> Result Fail
RError   --> Result Error
==============================

test_jsend.py
'''
import json
# from  .jsend import RSuccess, RFail, RError
from .jsend import *


def test_jsend():
    msg = make_successful_message()
    msg.status = STATUS_SUCCESSFUL
    msg.code = 100
    msg.data = {'hey': 'you, world, successful'}
    msg.desc = 'aloha'
    assert msg.is_successful()
    assert not msg.is_failed()
    assert msg.code == 100
    msg.code = 200
    assert msg.code == 200
    desc = 'desc changed'
    msg.desc = desc
    assert msg.desc == desc
    data = {'hey': 'what'}
    msg.data = data
    assert msg.data == data
    msg.status = STATUS_FAILED
    assert msg.status == STATUS_FAILED

    msg.status = STATUS_SUCCESSFUL
    msg.status = 'Any other status'
    assert msg.status == STATUS_SUCCESSFUL  # not change when setting invalid status

    code = 0
    data = {'body': 'way to go'}
    desc = 'description: test message'
    meta = ['127.0.0.1', 'google.com']
    d = {
        'status': 'S',
        'code': code,
        'data': data,
        'desc': desc,
        'meta': meta
    }
    msg = Message.load_from_dict(d)
    assert not msg.is_failed()
    assert msg.is_successful()
    assert msg.code == code
    assert msg.data == data
    assert msg.desc == desc
    assert msg.meta == meta

    msg = Message.loads(json.dumps(d))
    assert not msg.is_failed()
    assert msg.is_successful()
    assert msg.code == code
    assert msg.data == data
    assert msg.desc == desc
    assert msg.meta == meta
    d = msg.as_dict()
    for k, v in d.items():
        assert v == getattr(msg, k)
        assert v == msg[k]

    msg2 = Message.loads(msg.dumps())
    assert sorted([k, v] for k, v in msg.items()) == sorted([k, v] for k, v in msg2.items())

    msg2 = Message.loads(msg.dumps(skip_none=True))
    assert sorted([k, v] for k, v in msg.items()) == sorted([k, v] for k, v in msg2.items())
    assert 'code' in msg2
    msg2.code = None
    msg2.data = None
    msg2.desc = None
    msg2.meta =  None

    d3 = json.loads(msg2.dumps(skip_none=True))
    assert not 'code' in d3
    assert not 'data' in d3
    assert not 'desc' in d3
    assert not 'meta' in d3

    msg = make_successful_message(0, {'hello': 'world'}, 'description', 'meta')
    assert msg.is_successful()
    assert msg.code == 0

    msg = make_failed_message(0, {'hello': 'world'}, 'description', 'meta')
    assert msg.is_failed()
    assert msg.code == 0


if __name__ == '__main__':
    test_jsend()


jsend/jsend.py
# ** -- coding: utf-8 -- **
# !/usr/bin/env python
#
# Copyright (c) 2011 darkdarkfruit <darkdarkfruit@gmail.com>
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#

'''
compatable with jsend:
http://labs.omniti.com/labs/jsend

shortely:
=================================================================
A basic JSend-compliant response is as simple as this:
{{{
{
    status : "success",
    data : {
        "post" : { "id" : 1, "title" : "A blog post", "body" : "Some useful content" }
     }
}
}}}

When setting up a JSON API, you'll have all kinds of different types of calls and responses.
JSend separates responses into some basic types, and defines required and optional keys for each type:

| ''Type'' | ''Description''                                        | ''Required Keys'' | ''Optional Keys'' |
|----------+--------------------------------------------------------+-------------------+-------------------|
| success  | All went well, and (usually) some data was returned.   | status, data      |                   |
|          |                                                        |                   |                   |
| fail     | There was a problem with the data submitted,           | status, data      |                   |
|          | or some pre-condition of the API call wasn't satisfied |                   |                   |
|          |                                                        |                   |                   |
| error    | An error occurred in processing the request,           | status, message   | code, data        |
|          | i.e. an exception was thrown                           |                   |                   |

=================================================================


changelog: (2012-07-26)
In pracmatical usage, a key "message" for all responese is very convenient for development.
So add 'message' key to "Optional Keys" for type(success, fail)


| ''Type'' | ''Description''                                        | ''Required Keys''  | ''Optional Keys'' |
|----------+--------------------------------------------------------+--------------------+-------------------|
| success  | All went well, and (usually) some data was returned.   | status, code, data | message           |
|          |                                                        |                    |                   |
| fail     | There was a problem with the data submitted,           | status, code, data | message           |
|          | or some pre-condition of the API call wasn't satisfied |                    |                   |
|          |                                                        |                    |                   |
| error    | An error occurred in processing the request,           | status, message    | code, data        |
|          | i.e. an exception was thrown                           |                    |                   |





changelog: (2012-07-16)
In pracmatical usage, code representing status is very important, so add 'code' key to "Required Keys"


| ''Type'' | ''Description''                                        | ''Required Keys''  | ''Optional Keys'' |
|----------+--------------------------------------------------------+--------------------+-------------------|
| success  | All went well, and (usually) some data was returned.   | status, code, data |                   |
|          |                                                        |                    |                   |
| fail     | There was a problem with the data submitted,           | status, code, data |                   |
|          | or some pre-condition of the API call wasn't satisfied |                    |                   |
|          |                                                        |                    |                   |
| error    | An error occurred in processing the request,           | status, message    | code, data        |
|          | i.e. an exception was thrown                           |                    |                   |






this module include 3 classes:
==============================
RSuccess --> Result Success
RFail    --> Result Fail
RError   --> Result Error
==============================


'''

# __all__ = ['RSuccess', 'RFail', 'RError', 'jsend_parse']

from enum import Enum

import json

STATUS_SUCCESSFUL = 'S'
STATUS_FAILED = 'F'

STATUSES = (STATUS_SUCCESSFUL, STATUS_FAILED)


class Message(dict):
    """
    (status is always in state: "S" or "F"(represents "Successful", "Failed"), no 3th state)

    |--------+--------+----------+----------+-------------------------------------------------------|
    | Field  | type   | Required | Optional | Meaning                                               |
    |--------+--------+----------+----------+-------------------------------------------------------|
    | status | string | * (S/F)  |          | Is the response successful?                           |
    | code   | any    |          | *        | CODE for application logic(Normally it is an integer) |
    | data   | any    |          | *        | Data(payload) of the response                         |
    | desc   | any    |          | *        | Description: normally it's a helping infomation       |
    | meta   | any    |          | *        | eg: servers/ips chain in distributed env.             |
    |        |        |          |          |                                                       |
    |--------+--------+----------+----------+-------------------------------------------------------|


    """

    def __init__(self, status):
        super(Message, self).__init__()
        self['status'] = status if status in STATUSES else STATUS_FAILED
        self['code'] = None
        self['data'] = None
        self['desc'] = None
        self['meta'] = None

    @property
    def status(self):
        """get status

        Arguments:
        - `self`:
        """
        return self['status']

    @status.setter
    def status(self, status):
        """get status

        Arguments:
        - `self`:
        """
        if status not in STATUSES:
            print(f'Invalid status: {status}. Nothing changes. Valid status are: {STATUSES}')
            return
        self['status'] = status

    @property
    def code(self):
        """get code

        Arguments:
        - `self`:
        """
        return self['code']

    @code.setter
    def code(self, code):
        """ set code

        Arguments:
        - 'self':
        - 'code' : set code
        """
        self['code'] = code

    @property
    def data(self):
        """get data

        Arguments:
        - `self`:
        """
        return self['data']

    @data.setter
    def data(self, data):
        """ set data

        Arguments:
        - 'self':
        - 'data' : set data
        """
        self['data'] = data

    @property
    def desc(self):
        """get desc

        Arguments:
        - `self`:
        """
        return self['desc']

    @desc.setter
    def desc(self, desc):
        """ set desc

        Arguments:
        - 'self':
        - 'desc' : set desc
        """
        self['desc'] = desc

    @property
    def meta(self):
        """get meta

        Arguments:
        - `self`:
        """
        return self['meta']

    @meta.setter
    def meta(self, meta):
        """ set meta

        Arguments:
        - 'self':
        - 'meta' : set meta
        """
        self['meta'] = meta

    def is_successful(self):
        return self.status == STATUS_SUCCESSFUL

    def is_failed(self):
        return self.status == STATUS_FAILED

    def as_dict(self, skip_none=False):
        if skip_none:
            return {k: v for k, v in self.items() if v is not None}
        else:
            return {k: v for k, v in self.items()}

    def dumps(self, skip_none=False):
        return json.dumps(self.as_dict(skip_none=skip_none))

    # def __getattr__(self, name):
    #     try:
    #         return self[name]
    #     except KeyError as e:
    #         raise AttributeError(e)
    #
    # def __setattr__(self, name, value):
    #     self[name] = value

    @classmethod
    def load_from_dict(cls, d) -> 'Message':
        if 'status' not in d:
            msg = 'Could not found field:"status" in dict, returning None'
            print(msg)
            return None

        if d['status'] not in STATUSES:
            print(f'Invalid status. Got: {d["status"]}, but valid are: {STATUSES}')
            return None

        msg = cls(d['status'])
        for key in msg.keys():
            if key in d:
                msg[key] = d[key]
        return msg

    @classmethod
    def loads(cls, json_bytes):
        try:
            d = json.loads(json_bytes)
        except Exception as e:
            msg = 'Json could not load from bytes[:30]:"%s". Returning None. Exception: %s' % (json_bytes[:30], e)
            print(msg)
            return None

        return cls.load_from_dict(d)


class SuccessfulMessage(Message):
    def __init__(self):
        super(SuccessfulMessage, self).__init__(STATUS_SUCCESSFUL)


class FailedMessage(Message):
    def __init__(self):
        super(FailedMessage, self).__init__(STATUS_FAILED)


def make_successful_message(code=None, data=None, desc=None, meta=None):
    msg = Message(STATUS_SUCCESSFUL)
    msg.code = code
    msg.data = data
    msg.desc = desc
    msg.meta = meta
    return msg


def make_failed_message(code=None, data=None, desc=None, meta=None):
    msg = Message(STATUS_FAILED)
    msg.code = code
    msg.data = data
    msg.desc = desc
    msg.meta = meta
    return msg
